core/services/llo/observation_context.go

```diff
@@ -0,0 +1,193 @@
+	executions map[streams.Pipeline]*execution
+}
+
+func NewObservationContext(r Registry, t Telemeter) ObservationContext {
+	return newObservationContext(r, t)
+}
+
+func newObservationContext(r Registry, t Telemeter) *observationContext {
+	return &observationContext{r, t, sync.Mutex{}, make(map[streams.Pipeline]*execution)}
+}
+
+func (oc *observationContext) Observe(ctx context.Context, streamID streams.StreamID, opts llo.DSOpts) (val llo.StreamValue, err error) {
+	run, trrs, err := oc.run(ctx, streamID)
+	if err != nil {
+		// FIXME: This is a hack specific for V3 telemetry, future schemas should
+		// use a generic stream value telemetry instead
+		// https://smartcontract-it.atlassian.net/browse/MERC-6290
+		oc.t.EnqueueV3PremiumLegacy(run, trrs, streamID, opts, val, err)
+		return nil, err
+	}
+	// Extract stream value based on streamID attribute
+	for _, trr := range trrs {
+		if trr.Task.TaskStreamID() != nil && *trr.Task.TaskStreamID() == streamID {
+			val, err = resultToStreamValue(trr.Result.Value)
+			if err != nil {
+				return nil, fmt.Errorf("failed to convert result to StreamValue for streamID %d: %w", streamID, err)
+			}
+			return val, nil
+		}
+	}
+	// If no streamID attribute is found in the task results, then assume the
+	// final output is the stream ID and return that. This is safe to do since
+	// the registry will never return a spec that doesn't match either by tag
+	// or by spec streamID.
+
+	val, err = extractFinalResultAsStreamValue(trrs)
+	// FIXME: This is a hack specific for V3 telemetry, future schemas should
+	// use a generic stream value telemetry instead
+	// https://smartcontract-it.atlassian.net/browse/MERC-6290
+	oc.t.EnqueueV3PremiumLegacy(run, trrs, streamID, opts, val, err)
+	return
+}
+
+func resultToStreamValue(val interface{}) (llo.StreamValue, error) {
+	switch v := val.(type) {
+	case decimal.Decimal:
+		return llo.ToDecimal(v), nil
+	case float64:
+		return llo.ToDecimal(decimal.NewFromFloat(v)), nil
+	case pipeline.ObjectParam:
+		switch v.Type {
+		case pipeline.DecimalType:
+			return llo.ToDecimal(decimal.Decimal(v.DecimalValue)), nil
+		default:
+			return nil, fmt.Errorf("don't know how to convert pipeline.ObjectParam with type %d to llo.StreamValue", v.Type)
+		}
+	default:
+		return nil, fmt.Errorf("don't know how to convert pipeline output result of type %T to llo.StreamValue (got: %v)", val, val)
+	}
+}
+
+// extractFinalResultAsStreamValue extracts a final StreamValue from a TaskRunResults
+func extractFinalResultAsStreamValue(trrs pipeline.TaskRunResults) (llo.StreamValue, error) {
+	// pipeline.TaskRunResults comes ordered asc by index, this is guaranteed
+	// by the pipeline executor
+	finaltrrs := trrs.Terminals()
+
+	// HACK: Right now we rely on the number of outputs to determine whether
+	// its a Decimal or a Quote.
+	// This is a hack to support the legacy "Quote" case.
+	// Future stream specs should use streamID tags instead.
+	switch len(finaltrrs) {
+	case 1:
```

Comment:
Where do these case values come from? Why is case 2 skipped?

Reply:
Oh I see it's the number of results in the task run..Won't flexible schema result in a variable number of results - how would you know if it's benchmark/bid/ask vs bid/ask/marketDepth for example?  Or is a taskRunResult just one stream result, and one stream result always one datapoint?

Reply:
This is only needed for when you specify streamID at the job level, which this PR takes us away from. So we need to continue supporting this for the old use-case but for new cases its not relevant since the task can be tagged with a streamID